File: src/main/java/com/freya02/bot/utils/DecomposedName.java
package com.freya02.bot.utils;

import com.freya02.docs.DocSourceType;
import okhttp3.HttpUrl;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public record DecomposedName(@Nullable String packageName, @NotNull String className) {
	@NotNull
	public static String getSimpleClassName(@NotNull String fullName) {
		for (int i = 0; i < fullName.length() - 1; i++) {
			if (fullName.charAt(i) == '.' && Character.isUpperCase(fullName.charAt(i + 1))) {
				return fullName.substring(i + 1);
			}
		}

		//No package if naming conventions are respected
		return fullName;
	}

	public static @Nullable String getPackageName(@NotNull String fullName) {
		for (int i = 0; i < fullName.length() - 1; i++) {
			if (fullName.charAt(i) == '.' && Character.isUpperCase(fullName.charAt(i + 1))) {
				return fullName.substring(0, i);
			}
		}

		//No package if naming conventions are respected
		return null;
	}

	@Contract("_ -> new")
	@NotNull
	public static DecomposedName getDecomposition(@NotNull String fullName) {
		for (int i = 0; i < fullName.length() - 1; i++) {
			if (fullName.charAt(i) == '.' && Character.isUpperCase(fullName.charAt(i + 1))) {
				return new DecomposedName(fullName.substring(0, i), fullName.substring(i + 1));
			}
		}

		//No package if naming conventions are respected
		return new DecomposedName(null, fullName);
	}

	@Contract("_, _ -> new")
	@NotNull
	public static DecomposedName getDecompositionFromUrl(@NotNull DocSourceType sourceType, @NotNull String target) {
		final HttpUrl sourceUrl = HttpUrl.get(sourceType.getSourceUrl());
		final HttpUrl targetUrl = HttpUrl.get(target);

		final List<String> rightSegments = new ArrayList<>(targetUrl.pathSegments().subList(sourceUrl.pathSize(), targetUrl.pathSize()));

		//Remove java 9 modules from segments
		if (rightSegments.get(0).startsWith("java.")) rightSegments.remove(0);

		//All segments except last
		final List<String> packageSegments = rightSegments.subList(0, rightSegments.size() - 1);

		final String lastSegment = rightSegments.get(rightSegments.size() - 1);

		//Remove .html extension
		final String className = lastSegment.substring(0, lastSegment.lastIndexOf('.'));

		return new DecomposedName(String.join(".", packageSegments), className);
	}
}

File: doxxy-docs/src/main/kotlin/dev/freya02/doxxy/docs/utils/DecomposedName.kt
package dev.freya02.doxxy.docs.utils

import org.jsoup.nodes.Element

@JvmRecord
internal data class DecomposedName(val packageName: String?, val className: String) {
    companion object {
        fun getSimpleClassName(fullName: String): String {
            for (i in 0 until fullName.length - 1) {
                if (fullName[i] == '.' && Character.isUpperCase(fullName[i + 1])) {
                    return fullName.substring(i + 1)
                }
            }

            //No name if naming conventions aren't respected
            return fullName
        }

        fun getDecompositionFromLink(element: Element): DecomposedName {
            require(element.tag().name == "a")
            val href = element.attr("href").ifEmpty { error("href is missing in $element") }
            val title = element.attr("title").ifEmpty { error("title is missing in $element") }

            if (title.startsWith("type parameter"))
                return DecomposedName(null, element.ownText())

            val simpleClassName = href
                .substringAfterLast('/') // Last path segment
                .dropLast(5) // remove .html

            val packageName = title.substringAfterLast(' ')

            return DecomposedName(packageName, simpleClassName)
        }
    }
}